core/modules/parser/SelectParser.cc

```diff
@@ -74,13 +107,74 @@ namespace parser {
 ////////////////////////////////////////////////////////////////////////
 class AntlrParser {
 public:
-    AntlrParser(std::string const& q)
-        : statement(q),
-          stream(q, std::stringstream::in | std::stringstream::out),
-          lexer(stream),
-          parser(lexer)
-        {}
-    void run() {
+    virtual ~AntlrParser() {}
+    virtual void setup() = 0;
+    virtual void run() = 0;
+    virtual query::SelectStmt::Ptr getStatement() = 0;
+
+protected:
+    enum State {
+        INIT, SETUP_DONE, RUN_DONE
+    };
+    std::string stateString(State s) {
+        if (s == INIT) {
+            return "INIT";
+        } else if (s == SETUP_DONE) {
+            return "SETUP_DONE";
+        } else if (s == RUN_DONE) {
+            return "RUN_DONE";
+        }
+        return "??";
+    }
+
+    void changeState(State to) {
+        switch (_state) {
+            case INIT:
+                if (SETUP_DONE != to) {
+                    throw ParseException("Parse error(INTERNAL):invalid state transition from INIT to "
+                            + stateString(to));
+                }
+                _state = SETUP_DONE;
+                break;
+
+            case SETUP_DONE:
+                if (RUN_DONE != to) {
+                    throw ParseException("Parse error(INTERNAL):invalid state transition from SETUP_DONE to "
+                            + stateString(to));
+                }
+                _state = RUN_DONE;
+                break;
+
+            case RUN_DONE:
+                // there are no valid transitions from RUN_DONE
+                throw ParseException("Parse error(INTERNAL):invalid state transition from RUN_DONE to "
+                        + stateString(to));
+                break;
+        }
+    }
+
+    bool runTransitionDone() const { return _state == RUN_DONE; }
+
+private:
+    State _state {INIT};
+};
+
+class Antlr2Parser : public AntlrParser {
+public:
+    Antlr2Parser(std::string const& q)
+    : statement(q)
+    , stream(q, std::stringstream::in | std::stringstream::out)
+    , lexer(stream)
+    , parser(lexer)
+    {}
+
+    void setup() override {
```

Comment:
AFAK no performance benefits. This is just a way to tell readers of your code than this is the final class in a hierarchy. And secondly, this would prevent other developers from thoughtlessly overriding your class w/o diving into the implementation and trying to understand why you think you didn't want them to override it.
Otherwise, "override" is as fine as "final".